homebridge/thermostat.ts

```diff
@@ -0,0 +1,126 @@
+import { Logging, PlatformAccessory } from 'homebridge'
+
+import { RingDevice } from '../api'
+import { ThermostatMode } from '../api/ring-types'
+import { BaseDeviceAccessory } from './base-device-accessory'
+import { RingPlatformConfig } from './config'
+import { hap } from './hap'
+
+export class ExampleThermostatAccessory extends BaseDeviceAccessory {
+  constructor(
+    public readonly device: RingDevice,
+    public readonly accessory: PlatformAccessory,
+    public readonly logger: Logging,
+    public readonly config: RingPlatformConfig
+  ) {
+    super()
+
+    const { Characteristic, Service } = hap
+
+    // Required Characteristics
+
+    this.registerCharacteristic({
+      characteristicType: Characteristic.CurrentHeatingCoolingState,
+      serviceType: Service.Thermostat,
+      getValue: ({ mode }) => {
+        switch (mode) {
+          case 'off':
+            return Characteristic.CurrentHeatingCoolingState.OFF
+          case 'heat':
+            return Characteristic.CurrentHeatingCoolingState.HEAT
+          case 'cool':
+            return Characteristic.CurrentHeatingCoolingState.COOL
+        }
+      },
+    })
+
+    this.registerCharacteristic({
+      characteristicType: Characteristic.TargetHeatingCoolingState,
+      serviceType: Service.Thermostat,
+      getValue: ({ mode }) => {
+        switch (mode) {
+          case 'off':
+            return Characteristic.CurrentHeatingCoolingState.OFF
+          case 'heat':
+            return Characteristic.CurrentHeatingCoolingState.HEAT
+          case 'cool':
+            return Characteristic.CurrentHeatingCoolingState.COOL
+        }
+      },
+      setValue: (mode: ThermostatMode) => {
+        this.logger.info(`Setting ${this.device.name} mode to ${mode}`)
+
+        return this.device.setInfo({ device: { v1: { mode } } })
+      },
+    })
+
+    this.registerCharacteristic({
```

Comment:
@dgreif Do you have any guidance about when to use `registerCharacteristic` as opposed to `registerObservableCharacteristic`? I noticed `TemperatureSensor` [uses `registerObservableCharacteristic`](https://github.com/dgreif/ring/blob/d1cf0a5d3eae4d506d7aea839b4bb73140a042a0/homebridge/temperature-sensor.ts#L22) for the `CurrentTemperature` characteristic on the `TempSensorService` service.

Reply:
I'm hesitant to add the component devices to the flat data of the parent.  The downside is you don't have a reference to the `RingDevice` instance for that component device, just to its data.  It may be more useful to add a `getComponentDevices(parentZid: string) => RingDevice[]` method to each device.  I haven't seen the `addLinkedService` method before but it definitely seems useful in this case.  Maybe after all accessories have been created in the platform, we could then go back and do linking.

Reply:
Man, GitHub is being really odd about how it links my responses to your comments today 😕 .  Might want to have someone look into that 😉.  Anyways, `registerObservableCharacteristic` is part of the `BaseAccessory` and is intended to be used when you already have an `Observable` that is producing the exact values that you want to use for `Characteristic` updates.  It is mainly used by non-`RingDevice`s, like the camera and location Modes.  `registerCharacteristic` is on `BaseDataAccessory` and is very useful when you have a `RingDevice` and want to map from the `onData` events to the characteristic value.  In general, if you have a `RingDevice`, you should use `registerCharacteristic`.  The reason I did the `TemperatureSensor` and a few others differently is because I wanted to have a subscription to the `Observable` that would log when the sensor was faulted.  At that point I already had the `Observable` set up, so no need to do the `onData` mapping again by using `registerCharacteristic`.  Now that I look at it, I probably would have been better off adding a `handleUpdate` option to `registerCharacteristic` that would get called each time the characteristic changed value and could be used specifically for logging the events that we care about.